user_tools/custom_speedup_factors/generate_speedup_factors.py

```diff
@@ -55,13 +57,26 @@
     mapping_info = mapping_info.groupby(['SQL Node'])['Child Node'].apply(','.join).reset_index()
 
     # - process sql_plan_metrics_for_application.csv
-    #   - load in "duration" (CPU) or "op time" (GPU)
+    #   - load in "duration" (CPU)
     #   - replace WholeStageCodegen (CPU only) with list of operators from mapping lookup file
     #     - mapping_info.parent = sql_times.nodeName
     cpu_sql_info = pd.read_csv(cpu_dir + "/" + app + "/sql_plan_metrics_for_application.csv")
     cpu_sql_times = cpu_sql_info[cpu_sql_info["name"] == "duration"]
     cpu_sql_combined = cpu_sql_times.set_index('nodeName').join(mapping_info.set_index('SQL Node'), how='left')
 
+    #  - parse WholeStageCodegen durations with child node mapping
+    cpu_sql_times_df = cpu_sql_combined[['Child Node', 'max_value']]
+
+    for index, row in cpu_sql_times_df.iterrows():
+        operators = str(row['Child Node']).split(',')
+        duration = row['max_value']/len(operators)/1000.0
```

Comment:
Here division by 1000 is done so that the duration of CPU and GPU are in same metric? IIRC, duration is in `ms` and timing(GPU op time) is in `nanoSecond` .  Please ignore this comment  if 1000.0 is done for some other case.

Reply:
Yes, that is the intent.  I was assuming that the max_value of duration for CPU `sql_plan_metrics_for_application.csv` is in milliseconds where as value of `Stage Duration` in `sql_to_stage_information.csv` for both CPU and GPU is in nanoseconds.  Let me know if that assumption is wrong and if I need to update calculation logic.

Reply:
Note that if I don't divide by 1000, then it makes overall CPU durations really high with values from WholeStageCodegen and then speedup factors aren't reasonable at all.

Reply:
Thanks! I think it is okay to keep it as it is.